### PR TITLE
Feature #108675 feat: Rewrite hierarchy model, controller, views

### PR DESCRIPTION
**Pull request by pw4swapnil**
- [x] #108675: Rewrite hierarchy model, controller, view.
- [x] #108720: To add configuration setting
- [x] #108698: Backend >> Rewrite form view model
- [x] #109531: Backend >> Export CSV

### CODE REVIEW

File: administrator/access.xml [hunk 1]
```diff
@@ -8,13 +8,15 @@
 		<action name="core.edit" title="JACTION_EDIT" description="JACTION_EDIT_COMPONENT_DESC" />
 		<action name="core.edit.state" title="JACTION_EDITSTATE" description="JACTION_EDITSTATE_COMPONENT_DESC" />
 		<action name="core.edit.own" title="JACTION_EDITOWN" description="JACTION_EDITOWN_COMPONENT_DESC" />
+		<action name="core.edit.chart" title="COM_HIERARCHY_CHART_TITLE" description="COM_HIERARCHY_CHART_VIEW_DESC" />
```
Comment (manojLondhe) [inline administrator/access.xml]: these should be

```core.chart.view
core.csv.import
core.csv.export```

@pw4swapnil

File: administrator/controllers/hierarchys.php [hunk 2]
```diff
@@ -47,185 +48,131 @@ public function getModel($name = 'hierarchy', $prefix = 'HierarchyModel', $confi
...
 	 *
+	 * @param   string  $email  email.
+	 * 
 	 * @return  void
 	 *
-	 * @since   3.0
+	 * @since   1.0
 	 */
-	public function setUser()
+	public static function getUserId($email)
```
Comment (manojLondhe) [inline administrator/controllers/hierarchys.php]: why is this static?

why do you need query limit for load result?

File: administrator/hierarchy.php [hunk 3]
```diff
@@ -22,3 +22,7 @@
 $controller = JControllerLegacy::getInstance('Hierarchy');
 $controller->execute(JFactory::getApplication()->input->get('task'));
 $controller->redirect();
+
+// Initialize hierarchy js
```
Comment (manojLondhe) [inline administrator/hierarchy.php]: can you move this up?

File: administrator/language/en-GB/en-GB.com_hierarchy.ini [hunk 4]
```diff
@@ -31,8 +31,8 @@ COM_HIERARCHY_VIEW_FILE="[View File]"
 
 COM_HIERARCHY_TITLE_HIERARCHYS="Hierarchy"
 COM_HIERARCHY_HIERARCHYS_ID="ID"
-COM_HIERARCHY_HIERARCHYS_USER_ID="User Id"
-COM_HIERARCHY_HIERARCHYS_USER_NAME="User Name"
+COM_HIERARCHY_HIERARCHYS_USER_ID="Id"
```
Comment (manojLondhe) [inline administrator/language/en-GB/en-GB.com_hierarchy.ini]: Id => ID

File: administrator/models/fields/contextlist.php [hunk 5]
```diff
@@ -61,15 +61,27 @@ protected function getOptions()
...
 			{
-				$context   = $val->context;
+				$context   = $context->context;
 				$options[] = JHtml::_('select.option', $context, $context);
 			}
 		}
 
 		$options = array_unique($options, SORT_REGULAR);
 
+		foreach ($options as $elementKey => $element)
```
Comment (manojLondhe) [inline administrator/models/fields/contextlist.php]: add comments

File: administrator/models/hierarchys.php [hunk 6]
```diff
@@ -176,18 +188,16 @@ protected function getListQuery()
...
-			$deleteHierarchy = $db->getQuery(true);
-			$deleteHierarchy->delete($db->quoteName('#__hierarchy_users'));
-			$deleteHierarchy->where('id IN (' . $id . ')');
-			$db->setQuery($deleteHierarchy);
-			$confrim = $db->execute();
+			$deleteHierarchy = $this->db->getQuery(true);
+			$deleteHierarchy->delete($this->db->quoteName('#__hierarchy_users'));
+			$deleteHierarchy->where('user_id IN (' . $id . ')');
+			$this->db->setQuery($deleteHierarchy);
+			$confrim = $this->db->execute();
```
Comment (manojLondhe) [inline administrator/models/hierarchys.php]: `confrim` correct typo

File: media/samplecsv/userImport.csv [hunk 7]
```diff
@@ -1,3 +1,3 @@
-empId,bossId
-1,2
-3,2
+SubuserId,Name,Username,User_email,Id,User_id,Reports_to,Context,Context_id,Created_by,Modified_by,Created_date,Modified_date,State,Note,Reports_to_email
+310,SuperU,admin,swapnil_p@techjoomla.com,1325,310,312,,0,310,310,2017-11-03 06:46:58,2014-10-31 08:03:55,1,,Raviraj@mailinator.com
```
Comment (manojLondhe) [inline media/samplecsv/userImport.csv]: use dummy data

File: site/controllers/hierarchys.php [hunk 8]
```diff
@@ -1,49 +1,70 @@
...
-	public function save()
+
+	/**
+	 * Method to get users to manage hierarchy.
+	 *
+	 * @return  array
+	 *
+	 * @since    1.6
+	 */
+	public function getAlsoReportsTo()
```
Comment (manojLondhe) [inline site/controllers/hierarchys.php]: what is this needed for?
can you rename this ?

File: site/helpers/hierarchy.php [hunk 9]
```diff
@@ -16,248 +16,108 @@
...
 
-	public static function getTrainingAdminExit($uid = null)
-	{
-		$params = JComponentHelper::getParams('com_hierarchy');
-		$groupid = $params->get('groupid', '','INT');
-		$user = JFactory::getUser();
-		$uid = $user->id;
-
-		if ($uid)
+		if ($integration == "2")
```
Comment (manojLondhe) [inline site/helpers/hierarchy.php]: use tjsocial library

File: administrator/access.xml (hunk 1 above)
Comment (pw4swapnil) [reply administrator/access.xml]: done!

File: administrator/controllers/hierarchys.php (hunk 2 above)
Comment (pw4swapnil) [reply administrator/controllers/hierarchys.php]: No need to use create a static function, will be able to use it in future if needed somewhere else in the hierarchy. Removed static.
Removed query limit.

File: administrator/hierarchy.php (hunk 3 above)
Comment (pw4swapnil) [reply administrator/hierarchy.php]: done!

File: administrator/language/en-GB/en-GB.com_hierarchy.ini (hunk 4 above)
Comment (pw4swapnil) [reply administrator/language/en-GB/en-GB.com_hierarchy.ini]: done!

File: administrator/models/fields/contextlist.php (hunk 5 above)
Comment (pw4swapnil) [reply administrator/models/fields/contextlist.php]: done!

File: administrator/models/hierarchys.php (hunk 6 above)
Comment (pw4swapnil) [reply administrator/models/hierarchys.php]: done!

File: media/samplecsv/userImport.csv (hunk 7 above)
Comment (pw4swapnil) [reply media/samplecsv/userImport.csv]: done!

File: site/controllers/hierarchys.php (hunk 8 above)
Comment (pw4swapnil) [reply site/controllers/hierarchys.php]: done!
Removed this function and used "getReportsTo()".

**Pull request merged**